lib/api/bucketPutTagging.js

```diff
@@ -0,0 +1,104 @@
+const { waterfall } = require('async');
+const { Parser } = require('xml2js');
+const { errors } = require('arsenal');
+
+const collectCorsHeaders = require('../utilities/collectCorsHeaders');
+const { metadataValidateBucket } = require('../metadata/metadataUtils');
+const metadata = require('../metadata/wrapper');
+
+const tagKeyNotUnique = 'Cannot provide multiple Tags with the same key';
+const tagKeyInvalid = 'The TagKey you have provided is invalid';
+const tagValueInvalid = 'The TagValue you have provided is invalid';
+
+/**
+ * Format of xml request:
+
+ <Tagging xmlns="http://s3.amazonaws.com/doc/2006-03-01/">
+ <TagSet>
+ <Tag>
+ <Key>string</Key>
+ <Value>string</Value>
+ </Tag>
+ </TagSet>
+ </Tagging>
+ */
+
+function _parseXML(request, log, cb, parser) {
+    if (request.post === '') {
+        log.debug('request xml is missing');
+        return cb(errors.MalformedXML);
+    }
+    return parser.parseString(request.post, (err, result) => {
+        if (err) {
+            log.debug('request xml is malformed');
+            return cb(errors.MalformedXML);
+        }
+        return process.nextTick(() => cb(null, result));
+    });
+}
+
+function checkTags(tagSet, bucket, log, cb) {
+    if (tagSet === undefined) {
+        return cb(null, { tags: [], bucket });
+    }
+    const tags = Array.isArray(tagSet) ? tagSet : [tagSet];
+    const unique = [...new Set(tags.map(item => item.Key))];
+    let errMsg = null;
+    if (unique.length !== tags.length) {
+        errMsg = tagKeyNotUnique;
+    }
+    tags.forEach(item => {
+        if (item.Key.length > 128) {
+            errMsg = tagKeyInvalid;
+        }
+        if (item.Value.length > 256) {
+            errMsg = tagValueInvalid;
+        }
+    });
```

Comment:
As per this we check all the tagsets and return the last known error.
In AWS, I tried the same, it returns on the first error that was found.


For example, I tried the below JSON:
```json
{
    "TagSet": [
      {
        "Key": "organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1",
        "Value": "marketing"
      },
      {
        "Key": "o121",
        "Value": "rganization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1organization1"
      }
    ]
 }
```

Cloudserver (current PR code) gives me invalid value:
```sh
❯ aws s3api put-bucket-tagging --bucket test1 --tagging file://tagset.json --profile vault --endpoint-url http://localhost:8000

An error occurred (InvalidTag) when calling the PutBucketTagging operation: The TagValue you have provided is invalid
```

Where as AWS gives me invalid Tag Key:
```sh
❯ aws s3api put-bucket-tagging --bucket anuragplatform --tagging file://tagset.json

An error occurred (InvalidTag) when calling the PutBucketTagging operation: The TagKey you have provided is invalid
```

Reply:
See comment here for how I fixed error prioritisation https://github.com/scality/cloudserver/pull/4451#issuecomment-1108687268